### ISSUE TITLE
Interested in cooperation?

### ISSUE THREAD

**Issue by HDembinski** [dsaxton/resample#3]
tl;dr: I am a very experienced developer and a bootstrap expert, care to join forces?

I am the developer of [Boost.Histogram](https://github.com/boostorg/histogram) and the maintainer of [iminuit](https://github.com/scikit-hep/iminuit). I am a core member of the [Scikit-HEP community](https://scikit-hep.org). I am a senior developer in C++ and Python and a statistical expert. I read Efron's book "The introduction to the bootstrap" and a few others on resampling methods and think that resampling has a huge potential for my field (particle physics).

I think that we need a high-quality resampling package in Python, which ideally should become part of scipy at some point. I am privately using the balanced bootstrap since many years. I have an implementation of the balanced bootstrap in my tool library [pyik](https://github.com/hdembinski/pyik). I would like to move away from my own implementation, however, because that is not sustainable. I would be happy to support one of the projects that focus entirely on resampling, such as yours. Your package seems to be the most advanced in terms of quality and completeness from what I have seen on PyPI. Are you interested in collaboration?

Best regards!

**Comment by dsaxton** [comment]
Thanks for your interest @HDembinski, and I agree resampling methods don't seem terribly well supported in Python at the moment (that's what originally motivated me to make this package). I honestly haven't worked on this for some time now but it might be worth dusting off if there's interest. What did you have in mind in terms of collaboration?

**Comment by HDembinski** [comment]
Well, since you are the author, I thought you could suggest something.

A few things that strikes my eye:
* measure coverage and add more unit tests until line-coverage is 100 %
* expand the minimal documentation
* track performance with benchmarks and try to improve performance with numba

Bootstrapping is computationally intensive and high performance is important.

**Comment by HDembinski** [comment]
The items are listed in order of priority I would assign to them. As an outsider, I cannot know whether your package actually works. I need an extensive suite of unit tests and high coverage to trust this package. Once you have that, you can advertise it on the frontpage and gain more users.

To get more users, this package could become affiliated with Scikit-HEP, which gives this package a lot of visibility.

**Comment by HDembinski** [comment]
Another thing: you probably want to adapt to the new random number generator API in numpy, which finally moves away from the concept of using an implicit hidden global state for the random number generator. Users should be able to pass the seeded instance of their chosen random number generator to each function.

**Comment by dsaxton** [comment]
@HDembinski Yes, I agree testing is very important and we could do a lot to improve coverage. The non-deterministic pieces are a bit more challenging (e.g., how to test a function that purports to calculate a bootstrap confidence interval?), but it should be doable.

I like the benchmarking idea as well; for that we could likely set up some `asv` benchmarks: https://github.com/airspeed-velocity

I'd also like to add some type annotations to the existing functions (while not hugely important for a smallish project, it doesn't hurt and adds some clarity), and that should be pretty easy.

I'm also curious if you feel there are any features which are missing in your opinion? My original intention was to keep things very minimal (to simply provide a set of easy to use functions with few dependencies), but in so doing may have left some useful things out.

**Comment by HDembinski** [comment]
> The non-deterministic pieces are a bit more challenging (e.g., how to test a function that purports to calculate a bootstrap confidence interval?), but should be doable.

True. Unless you have a better idea, I think one should fix the random number generator and the seed for each test. A change in the implementation of, say, the bootstrap confidence interval, will more often make such a the test fail just because the random numbers are drawn in a different order, so one has to fix the expected value, but that is better than having no test.

> I like the benchmarking idea as well; for that we could likely set up some asv benchmarks: https://github.com/airspeed-velocity

I did not know this, but it looks great, indeed! Also great to advertise speed improvements.

> I'd also like to add some type annotations to the existing functions (while not hugely important for a smallish project, it doesn't hurt and adds some clarity), and that should be pretty easy.

Sounds good.

> I'm also curious if you feel there are any features which are missing in your opinion? My original intention was to keep things very minimal (to simply provide a set of easy to use functions with few dependencies), but in so doing may have left some useful things out.

> I'm also curious if you feel there are any features which are missing in your opinion? My original intention was to keep things very minimal (to simply provide a set of easy to use functions with few dependencies), but in so doing may have left some useful things out.

I am a big fan of minimal sufficient interfaces and small libraries that interoperate well with many others. To be honest, I haven't used your package a lot. I have my own implementation of a balanced bootstrap that I tested extensively and use for serious work, but like I said initially, this is not sustainable. I don't want to build my code into a successful professional package for resampling when you have done already so much in that regard.

One of the things I learned is that it is not enough to write brilliant code and offer it for free, you also need people to find it and grow a user base. This is not easy for OS devs that develop projects as a hobby.

That's why I suggested to join Scikit-HEP. I guess you are from a different field, perhaps even industry, but it would really help you to grow the user base to join some org. In high energy physics, we could really gain a lot from using bootstrapping more often. The technique is not well known in my field, but widely applicable to our statistical problems.

**Comment by dsaxton** [comment]
@HDembinski All sounds good to me, and I'm certainly open to including this package in the Scikit-HEP group if others feel it would make a good addition.

I'm going to see if I can clean this code up a little bit as well as adding better testing as you've suggested, and of course I would welcome any PRs (also feel free to post issues for things you'd like to see addressed).

**Comment by HDembinski** [comment]
Thanks for making me collaborator!

**Comment by HDembinski** [comment]
Can we connect outside of issues? How about Skype? My account is hans.dembinski.

**Comment by dsaxton** [comment]
> Can we connect outside of issues? How about Skype? My account is hans.dembinski.

Skype works
